### ISSUE TITLE
Support transparent enums with associated data

### ISSUE THREAD

**Issue by chinedufn** [chinedufn/swift-bridge#160]
Right now we support enums with unit variants, but we
have not yet implemented support for enums that have data.

```rust
// This is supported
enum ThisWorks {
    VariantA,
    VariantB,
}

// This is not yet supported
enum ThisDoesNotWorkYet {
    A(String),
    B {
        field: u32
    },
    C(Vec<u8>, i32),
    D {
        field1: u32,
        field2: SomeOpaqueRustType
    }
}
```

We'll need to add support for named and unnamed associaed data.

We do not need to support all of this in one pull request.

## Support Checklist

The following should be verified in tests:

- [x] Enum variant with one unnamed field
  - https://github.com/chinedufn/swift-bridge/pull/172

- [x] Enum variant with 2 unnamed fields
  - https://github.com/chinedufn/swift-bridge/pull/172

- [x] Enum variant with 1 named field
  - https://github.com/chinedufn/swift-bridge/pull/175

- [x] Enum variant with 2 unnamed field
  - https://github.com/chinedufn/swift-bridge/pull/175

### Contributing Guide

https://chinedufn.github.io/swift-bridge/contributing/adding-support-for-a-signature/index.html

### Transparent struct codegen and integration tests

Bridging enum variants should work similarly to bridging structs, so we may want to look over
the struct codegen and integration tests.

https://github.com/chinedufn/swift-bridge/blob/c1fd7177b5d3a8918e34b6bc96ae29437cfc1487/crates/swift-bridge-ir/src/codegen/codegen_tests/transparent_struct_codegen_tests.rs#L1-L704

https://github.com/chinedufn/swift-bridge/blob/b74432885120fbc6025349d8694505ec3b950918/crates/swift-integration-tests/src/shared_types/shared_struct.rs#L1-L80

https://github.com/chinedufn/swift-bridge/blob/b74432885120fbc6025349d8694505ec3b950918/SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/SharedStructTests.swift#L11-L45

### Enum codegen and integration tests

Here are the existing enum codegen and integration tests:

https://github.com/chinedufn/swift-bridge/blob/819e39160ec5b38dce2e60e2d39d078617ad158d/crates/swift-bridge-ir/src/codegen/codegen_tests/transparent_enum_codegen_tests.rs#L1-L348

https://github.com/chinedufn/swift-bridge/blob/3abac8c5488af967b83af21d924175fa393ab2dc/crates/swift-integration-tests/src/shared_types/shared_enum.rs#L1-L15

https://github.com/chinedufn/swift-bridge/blob/3abac8c5488af967b83af21d924175fa393ab2dc/SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/SharedEnumTests.swift#L11-L34

**Comment by NiwakaDev** [comment]
I want to try to implement this!

**Comment by chinedufn** [comment]
Woo! Go for it!

**Comment by NiwakaDev** [comment]
Hi, @chinedufn.

I have a question. What is this? I saw this in code-generation tests.
https://github.com/chinedufn/swift-bridge/blob/3238914429980d105d8d93de143d391a6f74ffa6/src/lib.rs#L49-L87

**Comment by chinedufn** [comment]
Hello there.

It's used to power the `#[swift_bridge(already_declared)]` attribute.

https://github.com/chinedufn/swift-bridge/blob/2bffcb08c1f0ca8bbcba9c78c5ae28a9cddfb632/crates/swift-bridge-ir/src/codegen/codegen_tests/already_declared_attribute_codegen_tests.rs#L149-L177

Just added a better comment https://github.com/chinedufn/swift-bridge/commit/e1b469dc2e6b5251f1313787e58b882c7fd61aaa

**Comment by NiwakaDev** [comment]
Hi, @chinedufn.

I have a question for the following code.

https://github.com/chinedufn/swift-bridge/blob/e5a0c893897f9aa2dddfdf1775876ff56b29d97c/crates/swift-bridge-ir/src/codegen/codegen_tests/transparent_enum_codegen_tests.rs#L24-L35

I guess that the two enum types seem same. 

Why does `swift-bridge` generate the two same enum types?

Maybe, same reason for #165 and #14?

**Comment by chinedufn** [comment]
They aren't always the same.

For example, if there is associated data it might look something like this:

```rust
 #[derive(Copy, Clone)] 
 pub enum SomeEnum { 
     Variant1(SomeRustType), 
     Variant2(u32)
 } 
  
 #[repr(C)] 
 #[doc(hidden)] 
 pub enum __swift_bridge__SomeEnum { 
     Variant1(*mut super::SomeRustType), 
     Variant2(u32)
 } 
```

---

But yeah, we technically don't need to have two enums when the FFI representation is the same as the main enum.

**Comment by NiwakaDev** [comment]
You mean `__swift_bridge__SomeEnum`  is an intermediate enum that bridges the enum type on the Rust side and that on the Swift side?

If so, I'll implement an intermediate enum based on [Enums With Fields](https://doc.rust-lang.org/reference/type-layout.html#reprc-enums-with-fields).

Something like: 

```rust
//Rust
#[repr(C)]
pub enum __swift_bridge__SomeEnum {
    A(i32),
    B(f32, u64)
}
```


```C
//header for Swift
//According to the above document, Rust`s SomeEnum is equal to this.

struct __swift_bridge__$SomeEnum {
    enum SomeEnumTag tag;
    union SomeEnumFields payload; 
};
```

We can use imported `union` on the Swift side based on [Using Imported C Structs and Unions in Swift](https://developer.apple.com/documentation/swift/using-imported-c-structs-and-unions-in-swift)

**Comment by chinedufn** [comment]
Yup `__swift_bridge__SomeEnum` is the C representation that we use for the enum. That's why it has the `#[repr(C)]` attribute.

---

Sounds good at a high level. Just need to make sure to convert the C `struct __swift_bridge__$SomeEnum` into a Swift enum.

As in

```rust
#[swift_bridge::bridge]
mod ffi {
    enum SomeEnum {
        Variant1(String)
    }

    extern "Rust" {
        fn foo() -> SomeEnum;
    }
}
```

Becomes on the Swift side

```swift
// Generated Swift (rough sketch)

enum SomeEnum {
    Variant1(String)
}

func foo () -> SomeEnum {
    /* Call __swift_bridge__$foo() and convert the value into `SomeEnum` */
}
````

**Comment by NiwakaDev** [comment]
I'll implement enum with named data this week.

**Comment by chinedufn** [comment]
Closed via #172 and #175
